### PR TITLE
Migrate customers listing

### PR DESCRIPTION
**Pull request by sarjon**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | This PR migrates customers list.
| Type?         | refacto
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Partially https://github.com/PrestaShop/PrestaShop/issues/10559
| How to test?  | Customers list and sorting/filtering should work as before. **NOTE: it does not migrate any actions, only list.** You can access new page by url `admin-dev/index.php/sell/customers/`

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/10916)

### CODE REVIEW

**Comment by matks** [comment]
@sarjon when you say:
> NOTE: it does not migrate any actions, only list

Do you mean "add/edit" page ? Or actions that can be triggered from the grid page ?

File: src/Core/Grid/Query/Builder/CustomerQueryBuilder.php [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+    }
+
+    private function appendTotalSpentQuery(QueryBuilder $queryBuilder)
+    {
+        $totalSpentQueryBuilder = $this->connection->createQueryBuilder()
+            ->select('SUM(total_paid_real / conversion_rate)')
+            ->from($this->dbPrefix . 'orders', 'o')
+            ->where('o.id_customer = c.id_customer')
+            ->andWhere('o.id_shop IN (:context_shop_ids)')
+            ->andWhere('o.valid = 1')
```
Comment (PierreRambaud) [inline src/Core/Grid/Query/Builder/CustomerQueryBuilder.php:138]: `o.valid is TRUE` => `'valid' => array('type' => self::TYPE_BOOL),`

File: src/Core/Grid/Query/Builder/CustomerQueryBuilder.php [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+                'gl',
+                'c.id_gender = gl.id_gender AND gl.id_lang = :context_lang_id'
+            )
+            ->leftJoin(
+                'c',
+                $this->dbPrefix . 'shop',
+                's',
+                'c.id_shop = s.id_shop'
+            )
+            ->where('c.deleted = 0')
```
Comment (PierreRambaud) [inline src/Core/Grid/Query/Builder/CustomerQueryBuilder.php:117]: `c.deleted is FALSE` => `'deleted' => array('type' => self::TYPE_BOOL, 'validate' => 'isBool', 'copy_post' => false),`

**Comment by sarjon** [comment]
@matks i mean actions that can be triggered from the grid, like bulk actions and toggle actions for some columns, they will be migrated in separate PR as commands. :)

File: src/Core/Grid/Query/Builder/CustomerQueryBuilder.php (hunk 1 above)
Comment (sarjon) [reply src/Core/Grid/Query/Builder/CustomerQueryBuilder.php:138]: i dont really get your comments. :/

Comment (PierreRambaud) [reply src/Core/Grid/Query/Builder/CustomerQueryBuilder.php:138]: it's a boolean not an integer. If we plan to move to another database like postgresql, nothing will work :P

Comment (sarjon) [reply src/Core/Grid/Query/Builder/CustomerQueryBuilder.php:138]: > we plan to move to another database like postgresql

we better dont. :D

i could do it with parameter binding, but is it worth?

Comment (PierreRambaud) [reply src/Core/Grid/Query/Builder/CustomerQueryBuilder.php:138]: When I said `move` i wanted to say `allow` my bad :P 
But if we have a boolean value we need to be consistent and use boolean in our requests.

File: src/PrestaShopBundle/Resources/views/Admin/TwigTemplateForm/bootstrap_4_layout.html.twig [hunk 3]
```diff
@@ -324,8 +324,8 @@
             <button type="reset"
                     name="{{ full_name }}[reset]"
                     class="btn btn-link js-reset-search btn d-block"
-                    data-url="{{ form.vars.attr['data-url'] }}"
-                    data-redirect="{{ form.vars.attr['data-redirect'] }}"
+                    data-url="{{ reset_url }}"
+                    data-redirect="{{ redirect_url }}"
             >
               <i class="material-icons">clear</i>
               {{ 'Reset'|trans({}, 'Admin.Global') }}
```
Comment (LouiseBonnard) [inline src/PrestaShopBundle/Resources/views/Admin/TwigTemplateForm/bootstrap_4_layout.html.twig]: It should be localized in _Admin.Actions_. :-)

**Comment by sarjon** [comment]
@LouiseBonnard fixed translations. @matks its ready for review. :)

File: src/Core/Grid/Query/Builder/CustomerQueryBuilder.php [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+     * @param Connection $connection
+     * @param string $dbPrefix
+     * @param int $contextLangId
+     * @param int[] $contextShopIds
+     */
+    public function __construct(
+        Connection $connection,
+        $dbPrefix,
+        $contextLangId,
+        $contextShopIds
```
Comment (tomas862) [inline src/Core/Grid/Query/Builder/CustomerQueryBuilder.php]: array $contextShopIds ?

File: src/Core/B2b/B2bFeature.php [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    {
+        $this->configuration->set('PS_B2B_ENABLE', 0);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function update($status)
+    {
+        $status ? $this->enable() : $this->disable();
```
Comment (matks) [inline src/Core/B2b/B2bFeature.php:87]: That's a funny use of ternary operator ^^ but it actually works and is very readable ... so 👍

**Review by matks — CHANGES_REQUESTED** [review]

**Comment by matks** [comment]
@sarjon I have the same feedback as Tomas 😉 

Otherwise ... this PR is just perfect 👏 !

File: src/Core/Grid/Column/Type/Common/BadgeColumn.php [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    protected function configureOptions(OptionsResolver $resolver)
+    {
+        parent::configureOptions($resolver);
+
+        $resolver
+            ->setRequired([
+                'field',
+            ])
+            ->setDefaults([
+                'badge_type' => 'success',
```
Comment (sarjon) [inline src/Core/Grid/Column/Type/Common/BadgeColumn.php:57]: @matks heres one more reason to have configurable templates for columns (and actions). Basically it's a simple `DataColumn` except it contains some styling options, so i had to create new type of column just to have different display of data. :/

File: src/Core/Grid/Query/Builder/CustomerQueryBuilder.php (hunk 4 above)
Comment (sarjon) [reply src/Core/Grid/Query/Builder/CustomerQueryBuilder.php]: done.

**Comment by sarjon** [comment]
@matks rebased and fixed comments. :+1:

**Review by matks — DISMISSED** [review]

File: src/Core/Grid/Column/Type/Common/BadgeColumn.php (hunk 6 above)
Comment (matks) [reply src/Core/Grid/Column/Type/Common/BadgeColumn.php:57]: If we enable this feature, we'll fix it, as written [here](https://github.com/PrestaShop/PrestaShop/issues/10910#issuecomment-430657534).

**Comment by marionf** [comment]
@sarjon 

I have 2 customers created on 2018-10-17

capture d ecran_456

If I search date from 2018-10-16 to 2018-10-17 I have no results, I should have two.

capture d ecran_455

Same thing if I search from 2018-10-17 to 2018-10-17 I have no results, I should have two.

capture d ecran_457

I should be able to filer by Newsletter and partner offer, there are not the arrows next to these fields

capture d ecran_458

**Comment by sarjon** [comment]
hi @marionf i did a small change regarding "Newsletter" and "Partner offers" filtering. See how "Enabled" field has dropdown with Yes/No options and is not sortable? so i did the same for other Yes/No columns to make it consistent, do you think i should revert it?

**Review by matks — DISMISSED** [review]

**Comment by marionf** [comment]
@sarjon Thank you for the fix on registration dates :+1: 

I talked with @TristanLDD for the 3 columns (enabled, newsletter & partner offers) we will keep the drop down and also add arrows next to these fields
We should have this behavior for all migrated pages 

Also, the icon of drop down for actions "display" and "delete" should be like this:

capture d ecran_473

Instead of this:

capture d ecran_471

We should also have this behavior for all migrated pages :)

**Comment by sarjon** [comment]
hey @marionf 

regarding "Enabled", "Newsletter" & "Partner offers", they are sortable now. :)

but for dropdown display, it affects all lists, so separate PR was made to implement that: https://github.com/PrestaShop/PrestaShop/pull/11095

**Comment by marionf** [comment]
Thank you @sarjon

**Review by matks — DISMISSED** [review]

File: src/PrestaShopBundle/Form/Admin/Type/SearchAndResetType.php [hunk 7]
```diff
@@ -52,7 +68,52 @@ public function buildView(FormView $view, FormInterface $form, array $options)
...
+            $resetUrl = $this->urlGenerator->generate(
+                $options['reset_route'],
+                $options['reset_route_params']
+            );
+        }
+
+        if (null !== $options['redirect_route']) {
+            $redirectUrl = $this->urlGenerator->generate(
+                $options['redirect_route'],
+                $options['redirect_route_params']
```
Comment (matks) [inline src/PrestaShopBundle/Form/Admin/Type/SearchAndResetType.php:84]: As said [here](https://github.com/PrestaShop/PrestaShop/pull/10839#discussion_r225958477) I'm a bit worried about these params 🤔

Comment (sarjon) [reply src/PrestaShopBundle/Form/Admin/Type/SearchAndResetType.php:84]: i think its okay to have array options for form types, as all forms are configured this way, wdyt?

**Comment by sarjon** [comment]
@matks rebased. :)

**Review by matks — APPROVED** [review]

**Comment by PierreRambaud** [comment]
Thanks @sarjon

**Pull request merged**